csharp/src/Ice/LocatorDiscovery/Locator.cs

```diff
@@ -271,45 +271,57 @@ internal void FoundLocator(ILocatorPrx locator)
                 _lookup.Communicator.Logger.Trace(_lookupTraceCategory, sb.ToString());
             }
 
-            int failureCount = 0;
+            // We retry only when at least one send succeeds and we don't get any reply.
+            // TODO: this _retryCount is really an attempt count not a retry count.
             for (int i = 0; i < _retryCount; ++i)
             {
+                var invocationTasks = new List<Task>(_lookups.Count);
                 foreach ((ILookupPrx lookup, ILookupReplyPrx lookupReply) in _lookups)
                 {
                     try
                     {
-                        await lookup.FindLocatorAsync(_instanceName, lookupReply).ConfigureAwait(false);
+                        invocationTasks.Add(lookup.FindLocatorAsync(_instanceName, lookupReply));
                     }
                     catch (Exception ex)
                     {
-                        lock (_mutex)
+                        invocationTasks.Add(Task.FromException(ex));
+                    }
+                }
+
+                Task t = Task.WhenAll(invocationTasks);
```

Comment:
Could this potentially hang if the user has specified a tcp lookup endpoint. I know they're not supposed to, but I'm not sure that we check. Is it otherwise possible for UDP to "hang" here?

Reply:
The lookup proxy is created as a datagram proxy:
```
var lookupPrx = ILookupPrx.Parse($"IceLocatorDiscovery/Lookup -d:{lookupEndpoints}", _communicator);
lookupPrx = lookupPrx.Clone(clearRouter: false);
```

The user cannot override this code by specifying tcp endpoints.

We also have a:
```
 Debug.Assert(endpoint.Transport == Transport.UDP);
```
Which could be replaced by a throw InvalidConfigurationException is you're concerned about bogus configuration.

So it's definitely datagram/UDP. 

I guess the sending with UDP can take a little time. Since we're not setting an explicit invocation timeout, we're using the default invocation timeout, which is by default 60s. Maybe we should use the plugin's timeout (by default 300ms) as invocation timeout on these lookup proxies?

Reply:
> Which could be replaced by a throw InvalidConfigurationException is you're concerned about bogus configuration.

Yes. I think it would indeed be better to throw an `InvalidConfigurationException` here instead.

Reply:
> Maybe we should use the plugin's timeout (by default 300ms) as invocation timeout on these lookup proxies?

Sure, that's probably better than 60s